docs/specs/add-kubernetes-support.md

```diff
@@ -0,0 +1,247 @@
+Kubernetes is the upstream community for OpenShift and the broker should be
+consumuing the latest upstream code for development purposes.
+
+This change will improve the Broker development process by having the Broker
+constantly being developed and tested against the latest upstream code.
+
+The Broker's functionality won't change, but all the tooling around it will.
+
+## Split Out All Cluster logic
+OpenShift is Kubernetes plus a few features, API name changes, and tools.
+In the core ansible-service-broker code, ```pkg/cluster/kubernetes.go```
+should be able to account for about 80% of the cluster opertations required
+by the broker. The remaining ~20% will be held in ```pkg/cluster/openshift.go```.
+
+### Code
+Start by creating a new pkg for cluster objects, ```pkg/cluster/...```. From
+there, the cluster pkg will hold ```pkg/cluster/cluster.go```,
+```pkg/cluster/kubernetes.go```, and ```pkg/cluster/openshift.go```.
+
+The file ```pkg/cluster/cluster.go```  will provide an object that will abstract
+cluster logic.
+
+Using an abstration is an advantage because the broker code will only have a
+single path. It will call the cluster object asking for an operation. For
+example, to create a new Pod:
+```diff
+- k8scli.CoreV1().Pods(ns).Create(pod)
++ Cluster.CreatePod(name, image, extraVars, pullPolicy, serviceAccountName, ns)
+```
+
+The Cluster object will recieve the request and call the correct API.
+```diff
++ func CreatePod(name string, image spec.Image, extraVars string,
++                pullPolicy string, serviceAccountName string, ns string) {
++	pod := &v1.Pod{
++		ObjectMeta: metav1.ObjectMeta{
++			Name: apbID,
++		},
++		Spec: v1.PodSpec{
++			Containers: []v1.Container{
++				{
++					Name:  "apb",
++					Image: spec.Image,
++					Args: []string{
++						action,
++						"--extra-vars",
++						extraVars,
++					},
++					ImagePullPolicy: pullPolicy,
++				},
++			},
++			RestartPolicy:      v1.RestartPolicyNever,
++			ServiceAccountName: serviceAccountName,
++		},
++	}
++
++       k8scli.CoreV1().Pods(ns).Create(pod)
++
++ }
+
+```
+
+If there are identical API objects between COEs, check which cluster is
+being used.
+```diff
++ func CreateClusterRoleBinding(roleBindingName string, svcAccountName string,
++                               ApbRole string, ns string) {
++	roleBindingM := map[string]interface{}{
++		"apiVersion": "v1",
++		"kind":       "RoleBinding",
++		"metadata": map[string]string{
++			"name":      roleBindingName,
++			"namespace": ns,
++		},
++		"subjects": []map[string]string{
++			map[string]string{
++				"kind":      "ServiceAccount",
++				"name":      svcAccountName,
++				"namespace": ns,
++			},
++		},
++		"roleRef": map[string]string{
++			"name": ApbRole,
++		},
++      }
++
++      if OpenShift {
++         OCcli.CoreV1().ClusterRoleBinding(ns).Create(roleBindingM)
++      } elif Kubernetes {
++         k8scli.CoreV1().ClusterRoleBinding(ns).Create(roleBindingM)
```

Comment:
Is there a mechanism in Go where we can have the OCP Object reuse the K8S implementation and just override/implement the deltas?

Opposed to needing to redeclare, if OCP/if K8S workflows

Assuming OCP workflow can reuse k8scli except for those cases where there are deltas in behavior.